tmt/steps/prepare/feature.py

```diff
@@ -0,0 +1,197 @@
+import dataclasses
+import re
+from typing import List, Optional
+
+import tmt
+import tmt.base
+import tmt.log
+import tmt.options
+import tmt.steps
+import tmt.steps.prepare
+import tmt.utils
+from tmt.steps.provision import Guest
+from tmt.utils import Command, field
+
+
+class Feature(tmt.utils.Common):
+    """ Base class for feature implementations """
+
+    KEY: str
+
+    def __init__(
+            self,
+            *,
+            parent: tmt.steps.prepare.PreparePlugin,
```

Comment:
This one can be even stricter, `'PrepareFeature'` - not the single quotes, by wrapping a type name with them you can use it in annotations before the type has been defined.

Reply:
Hi @happz,  if we use `parent: tmt.steps.prepare.PrepareFeature`  (as class `PrepareFeature` is defined in `feature.py`),  

```patch
--- a/tmt/steps/prepare/feature.py
+++ b/tmt/steps/prepare/feature.py
@@ -21,7 +21,7 @@ class Feature(tmt.utils.Common):
     def __init__(
             self,
             *,
-            parent: tmt.steps.prepare.PreparePlugin,
+            parent: tmt.steps.prepare.PrepareFeature,
             guest: Guest,
...<snip>...
```

an error in the following is hit,

```
Traceback (most recent call last):
  File "/home/huanli/.virtualenvs/tmt/bin/tmt", line 12, in <module>
    import tmt.cli
  File "/home/huanli/.virtualenvs/tmt/lib/python3.10/site-packages/tmt/cli.py", line 36, in <module>
    tmt.plugins.explore(tmt.log.Logger.get_bootstrap_logger())
  File "/home/huanli/.virtualenvs/tmt/lib/python3.10/site-packages/tmt/plugins/__init__.py", line 178, in explore
    _explore_packages(logger)
  File "/home/huanli/.virtualenvs/tmt/lib/python3.10/site-packages/tmt/plugins/__init__.py", line 145, in _explore_packages
    _explore_package(name, _TMT_ROOT / path, logger.descend())
  File "/home/huanli/.virtualenvs/tmt/lib/python3.10/site-packages/tmt/plugins/__init__.py", line 83, in _explore_package
    import_(module=f'{package}.{module}', logger=logger)
  File "/home/huanli/.virtualenvs/tmt/lib/python3.10/site-packages/tmt/plugins/__init__.py", line 193, in import_
    importlib.import_module(module)
  File "/usr/lib64/python3.10/importlib/__init__.py", line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "/home/huanli/.virtualenvs/tmt/lib/python3.10/site-packages/tmt/steps/prepare/feature.py", line 16, in <module>
    class Feature(tmt.utils.Common):
  File "/home/huanli/.virtualenvs/tmt/lib/python3.10/site-packages/tmt/steps/prepare/feature.py", line 24, in Feature
    parent: tmt.steps.prepare.PrepareFeature,
AttributeError: module 'tmt.steps.prepare' has no attribute 'PrepareFeature'
```

Any idea?

Reply:
Yes, do not use the full module path, it's defined in this file, isn't it? `parent: 'PrepareFeature'` should be perfectly fine.

Reply:
Thanks very much, `parent: 'PrepareFeature'`  works perfectly fine :-)